src/realm/sync/noinst/protocol_codec.hpp

```diff
@@ -491,6 +494,80 @@ class ClientProtocol {
         return action::ApplicationBug;
     }
 
+    template <typename Connection>
+    void parse_log_message(Connection& connection, HeaderLineParser& msg)
+    {
+        auto report_error = [&](const auto fmt, auto&&... args) {
+            auto msg = util::format(fmt, std::forward<decltype(args)>(args)...);
+            connection.handle_protocol_error(Status{ErrorCodes::SyncProtocolInvariantFailed, std::move(msg)});
+        };
+
+        auto session_ident = msg.read_next<session_ident_type>();
+        auto message_length = msg.read_next<size_t>('\n');
+        auto message_body_str = msg.read_sized_data<std::string_view>(message_length);
+        nlohmann::json message_body;
+        try {
+            message_body = nlohmann::json::parse(message_body_str);
+        }
+        catch (const nlohmann::json::exception& e) {
+            return report_error("Malformed json in log_message message: \"%1\": %2", message_body_str, e.what());
+        }
+        static const std::unordered_map<std::string_view, util::Logger::Level> name_to_level = {
+            {"fatal", util::Logger::Level::fatal},   {"error", util::Logger::Level::error},
+            {"warn", util::Logger::Level::warn},     {"info", util::Logger::Level::info},
+            {"detail", util::Logger::Level::detail}, {"debug", util::Logger::Level::debug},
+            {"trace", util::Logger::Level::trace},
+        };
+
+        // connection id (co_id) is optional and level and msg are optional if a co_id is provided
```

Comment:
i think we can loosen these invariants for simplicity

the server always has access to `co_id`, so it will likely always include it
we can make `level` optional and always default to debug/info (although the server would likely always include this as well)
and `msg` can also be "optional" and default to `<no message>` or something if the key is not provided (again, this would likely never happen)

lmk what you think